2.0.0

dcousens opened the pull request:
Resolves all remaining points in https://github.com/bitcoinjs/bip38/issues/5.
Specifically:
- consider removing coinstring [DONE]
- consider removing class instantiation, ~~pass script params as optional param~~ [DONE]

Script params were not put as optional, as there has been at least some consensus over time that the parameters given are suitable.
Happy to revise that.

I did not only keep the raw methods, as I felt the base58check encoded BIP38 string was important and specific to the encoding of this module.

Instead, I remove encoding/decoding of WIF format private keys as part of the `encrypt` interface.  You now specifically pass in a private key 32-byte buffer with a compression flag.

dcousens commented on index.js:
```diff
@@ -10,146 +10,113 @@ var curve = ecurve.getCurveByName('secp256k1')
+  result[1] = 0x42
+  result[2] = compressed ? 0xe0 : 0xc0
+  salt.copy(result, 3)
+  cipherText.copy(result, 7)
```
Optimized this a little bit, allocate the buffer all at once.

dcousens commented on test/index.js:
```diff
@@ -1,21 +1,20 @@
-        assert.equal(bip38.decrypt(f.bip38, f.passphrase), f.wif)
+        var result = bip38.decrypt(f.bip38, f.passphrase)
+
+        assert.equal(wif.encode(0x80, result.d, result.compressed), f.wif)
```
This is about the only real pain point IMHO,  but,  it should be simple for most users to navigate.
Thoughts @jprichardson ?

jprichardson commented:
Why remove the instantiation? I don't think I'm in favor of that unless a good case could be made.

dcousens commented:
Because if I want a parameter containing closure,  I can really easily make it.
This gives both freedoms,  without enforcing me to do `new` every call.

dcousens commented:
Because if I want a parameter containing closure,  I can really easily make it.

**In any case**,  with this PR,  what state would it contain?
All stateful parameters were removed,  they don't even exist in the proposed interface.

* `scryptParams` is now constant.
* `versions` were removed.

Even if they did exist,  this PR would give MORE freedom of integration,  as it doesn't force me to do `new` every call.

jprichardson commented:
> It is an implicit contract of state for something that isn't stateful.

But there is state, namely the scrypt parameters.

dcousens commented:
@jprichardson they are not state,  they are parameters.
State is variable.

If this were a `React` component, would you be putting `scryptParams` in `this.state` or `this.props`?

dcousens commented:
Any consensus on this?  ping @jprichardson ?

jprichardson commented:
I'll follow-up tomorrow or Wed...

On Mon, Aug 31, 2015 at 8:02 AM, Daniel Cousens <notifications@github.com>
wrote:

> Any consensus on this? ping @jprichardson
> <https://github.com/jprichardson> ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitcoinjs/bip38/pull/13#issuecomment-136364655>.
>



-- 
Simple & Secure Bitcoin Wallet: https://www.coinbolt.com
Bitcoin / JavaScript: http://cryptocoinjs.com
Follow JP Richardson on Twitter: https://twitter.com/jprichardson

jprichardson commented:
Here's what I'd accept...

A factory method called `create` that takes as input the scrypt params. It would then return an instance of the `Bip38` object. I'd also accept static methods that use the last param as scrypt params and default to the bip38 params.

ES2015 partial code: (not suggesting that it has to be ES2015, it's just more concise to communicate)

```js

class Bip38 {
  constructor (scryptParams) { ... }

  encrypt (...) { ... }

  decrypt (...) { ... }

  static encrypt (..., scryptParams) { ... }
  
  static decrypt (..., scryptParams) { ... }

  static create (scryptParams) { ... }
}
```

If you have a better suggestion, please use some code to communicate for the sake of clarity. Thanks.

dcousens commented:
> I'd also accept static methods that use the last param as scrypt params and default to the bip38 params.

That is what we have now?  No? Granted,  we just have to add that param to the end of the methods.

dcousens commented:
@jprichardson what is the difference between this and a synchronous `pbkdf2`?   Ignoring that `pbkdf2` is a one way function.

jprichardson commented:
> Conceptually, its the same. Yet in 1 case you're presumably happy with a free function, the other, an object.

I'm really warming up to the idea of just providing the functions. Especially since I can make it look really clean with ES6 now...

dcousens commented:
Any further thoughts on this @jprichardson ?

jprichardson commented:
I'm alright with this now, but I'm gonna do it in ES6 (Babel).

dcousens commented:
maybe merge this then do your changes @jprichardson ?

jprichardson commented:
> I'll rebase if necessary

Would be awesome.

dcousens commented:
Rebased and tested.

Note, there are no tests for the `progressCallback`s, so, they should be added/tested before release (if you get that far quickly).

jprichardson commented:
> Rebased and tested.

Awesome, I'm not satisfied with some of the dependencies just yet, addressing them on one by one basis. To start: https://github.com/bitcoinjs/wif/issues/5

dcousens commented:
@jprichardson `wif` has been updated, any other changes you want before I go ahead and rebase/commit the changes?

jprichardson commented:
Will review tomorrow.

dcousens commented:
@jprichardson yay

jprichardson commented:
Alright, almost all of it looks good. Some thoughts:

1. Documentation is not updated to reflect changes.
2. Returning `d` provides zero context to the end user unless you're familiar with ECC/ECDSA. Open to something else though.
3. Also function parameters named `string` or `buffer` only provide type information, but provide no semantic context. I think it's friendlier to the end user to have descriptive names. Could use `assert` to convey types. Or if you really wanted, append the type at the end of the name (yuck).

We get those patched up, I'll be willing and ready to merge :)

dcousens commented:
> Also function parameters named string or buffer only provide type information, but provide no semantic context. I think it's friendlier to the end user to have descriptive names.

You're decrypting a string to a private key.
IMHO that is what is relevant?  The type is all I'm passing IMHO.
I guess I don't know what other semantic context I'd give it...  "candidate"?

dcousens commented:
`wifString`?

dcousens commented:
Oh god lets finish this haha

jprichardson commented:
haha, alright, I'll tackle this bad boy on Monday.

jprichardson commented:
Why are the tests failing?

dcousens commented:
@jprichardson might be related to pbkdf2 tests failing.

dcousens commented:
I know its low priority.
But having `coinstring` still get pulled into my projects is odd just because of this module :smiley:

dcousens commented on index.js:
```diff
@@ -10,146 +10,137 @@ var curve = ecurve.getCurveByName('secp256k1')
+//   var d = BigInteger.fromBuffer(privateKey)
+//   var address = getAddress(d)
+//   var checksum = hash256(address).slice(0, 4)
+//   assert.deepEqual(salt, checksum)
```
Left out for now, as it instantly fails the tests... I'm not sure why.  But it is **unrelated** to this PR,  as the tests failed previously if this was added.

Lets get to that immediately after this :+1:

dcousens commented:
@jprichardson squashed and rebased.
This is good to go and very quick to review. :+1:

jprichardson commented:
Why the failing tests?

dcousens commented:
@jprichardson pushed up the missing compression flags in 63d1d4e.
They were why the the verify wasn't working ... yay :+1:

dcousens commented:
@jprichardson `0.10` is failing as it was dropped by the underlying crypto packages,  so I've removed it from travis.
I also removed `4.2`? Not sure why it is there.

dcousens commented:
https://travis-ci.org/bitcoinjs/bip38/jobs/180905933 is timing out,  restarted it and hopefully it passes.

dcousens commented:
Increased the timeout, and its all good now! :dancer:

jprichardson commented:
Awesome, thanks for your work on this.

1. You murdered the docs. They didn't even make it to the potty training stage. Poor docs.
2. https://github.com/bitcoin/bips/blob/master/bip-0038.mediawiki#discussion-item-scrypt-parameters - we gotta have configurable Scrypt params. Can be an optional param.

Thoughts?

dcousens commented:
@jprichardson if the parameters are changed,  won't that stop people being able to decrypt unless they *know* those parameters?

dcousens commented on README.md:
```diff
@@ -26,116 +22,49 @@ Package Info
-A field that accepts an object with the follow properties: `N`, `r`, and `p` to control the [scrypt](https://github.com/cryptocoinjs/scryptsy). The
-BIP38 standard suggests `N = 16384`, `r = 8`, and `p = 8`. However, this may yield unacceptable performance on a mobile phone. If you alter these parameters, it wouldn't be wise to suggest to your users that your import/export encrypted keys are BIP38 compatible. If you do, you may want to alert them of your parameter changes.
+### API
+### encrypt(buffer, compressed, passphrase, progressCallback)
```
What'd I miss?

dcousens commented on README.md:
````diff
@@ -26,116 +22,49 @@ Package Info
-// =>  '5KN7MzqK5wt2TP1fQCYyHBtDrXdJuXbUzm4A9rKAteGu3Qi5CVR'
-```
-
-**note:** To check for an invalid password, you'll want to generate the public address from the output of the `decrypt()` function. If it doesn't equal the expected address or the address checksum, then chances are, it's an invalid password. The reason that this logic was not included is because it would have required a lot of dependencies: `ECKey` and `Address`. Currently, `ECKey` is pretty heav ...
````
This is no longer relevant

dcousens commented on README.md:
```diff
@@ -26,116 +22,49 @@ Package Info
 
-A method that decrypts the encrypted string. `encryptedKey` is the string value of the encrypted key. `passphrase` is the passphrase to decrypt the key with.
-`progressCallback` is a function that receives an object in the form of: 
-{current: 1000, total: 262144, percent: 0.3814697265625}
```
Is anything other than `percent` worth it?

dcousens commented on README.md:
```diff
@@ -26,116 +22,49 @@ Package Info
-	private: 0x80, 
-  public: 0x0
-}
-bip38.encrypt(privateKeyWif, "super-secret", "1Jq6MksXQVWzrznvZzxkV6oY57oWXD9TXB"})
```
As mentioned,  the `version` aspect of the API currently doesn't exist

dcousens commented on README.md:
```diff
@@ -26,116 +22,49 @@ Package Info
-### scryptParams
-
-A field that accepts an object with the follow properties: `N`, `r`, and `p` to control the [scrypt](https://github.com/cryptocoinjs/scryptsy). The
-BIP38 standard suggests `N = 16384`, `r = 8`, and `p = 8`. However, this may yield unacceptable performance on a mobile phone. If you alter these parameters, it wouldn't be wise to suggest to your users that your import/export encrypted keys are BIP38 compatible. If you do, you may want to alert them of your parameter changes.
```
I guess I want to ask,  is *anyone* doing this?
I'm not against making them a global const, change-able,  but,   not part of the "public API" as such.

jprichardson replied:
> @jprichardson if the parameters are changed, won't that stop people being able to decrypt unless they know those parameters?

Of course. But that doesn't mean that we shouldn't prevent the option this case. I should state that I'm a person who usually hates options too. But the standard doesn't define the params. 

The standard should have had the params encoded since they're not fixed.

> I guess I want to ask, is anyone doing this?

I don't know.

> I'm not against making them a static variable, can be changed, but, not part of the "public API" as such.

Why the aversion to just adding an optional param?

dcousens replied:
> Why the aversion to just adding an optional param?

Just one more thing to have to put in the docs I guess.

jprichardson replied:
> Just one more thing to have to put in the docs I guess.

If you put the param there, I'll be happy and we can get 2.0 out the door. We can revisit the docs later <--- famous last words 😛

dcousens commented:
ping @jprichardson for 5d5a3d0

dcousens commented:
ping @jprichardson,  this should be good to go

jprichardson commented:
> edit: although the scrypt params option isn't documented... (I promise if you merge, i'll add it straight after)

👍
